python/lsst/obs/metadata/translator.py

```diff
@@ -0,0 +1,252 @@
+# This file is part of obs_metadata.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Classes and support code for metadata translation"""
+
+__all__ = ("MetadataTranslator",)
+
+from abc import abstractmethod
```

Comment:
I was following @TallJimbo's example from butler and I get the impression that there is a preference for `@abstractmethod` over `@abc.abstractmethod`.

Reply:
This has not been formalized anywhere, so this is *just* personal preference, but I have indeed recently really come to like `from <module> import <symbols>` style - I think maybe I picked it up from @pschella's daf_butler code (I would have guessed I got it from @timj first, but he just blamed me).  I think it *usually* makes localized logic easier to read by reducing clutter (at the expense of rare but admittedly annoying "where the heck did that symbol come from?!" questions that you then have to scroll up to answer), and it *really* makes refactoring symbols across package boundaries easier, in large part because it helps you see in one place at the top what a module's dependencies really are.

Reply:
It forces you to be explicit about what is used, and indeed helps refactoring, so I generally prefer `from <module> import <symbols>` to `import <module>`. Except for common cases (e.g. `numpy`, `scipy`, etc.) where `import <module> (as ...)` is easier to read, more Pythonic, or otherwise important (i.e. good to know, on the spot, if `sin` comes from `numpy` or `math`).
Yes, namespaces are one honking great idea, but so is moving things from one namespace to another (i.e. importing them into the local one). As long as the are not then exported again.

Reply:
I absolutely agree that keeping namespaces with symbols isn't *always* clutter; I'd say that it's clutter only when *both*:
 - you need to know where a symbol is from in order to know what it does (either because it's name is too short, it's fundamentally complex and hard to name, or you happen to care about the details of what its arguments are because you're e.g. debugging);
 - you don't already know where it's from.

Both of those are subjective.  I do think it's reasonable to expect most readers to know where this particular symbol comes from (and those don't can fall back to looking at the top, an important distinction from `from module import *`).  But my preference for this style is much more about code importing LSST modules, which typically have a large chunk of their namespace (e.g. "lsst.afw" or "lsst.meas") that conveys very little information and is never necessary to disambiguate symbols (e.g. there's only one `SourceCatalog` and one `Mask` across all packages).